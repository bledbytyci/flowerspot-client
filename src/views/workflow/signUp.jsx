import AuthActionCreators from '../../auth/authActionCreators.js';
import AuthSelectors from '../../auth/authSelectors.js';
import Modal from '../common/modal.jsx';
import PropTypes from 'prop-types';
import SignUpForm from './forms/user/signUpForm.jsx';
import SignUpSuccessForm from './forms/user/signUpSuccessForm.jsx';
import User from '../../user/user.js';
import validateSignUpForm from './forms/user/signUpFormValidation.js';
import ValidationModel from '../../setup/validationModel.js';
import ValidationSummary from '../common/validationSummary.jsx';
import React, {Component} from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import '../../styles/modal.css';

const mapStoreToProps = store => {
	return {
		isCreating: AuthSelectors.isCreating(store),
		validationModel: AuthSelectors.getValidationModel(store)
	}
}

const mapDispatchToProps = dispatch => {
	return {
		signUpUser: (user) => dispatch(AuthActionCreators.signUpUser(user))
	}
}

export class SignUp extends Component {
	constructor(props){
		super(props);

		this.state = {
			user: new User(),
			validationModel: new ValidationModel(),
			showSignUpSuccessModal: false,
			showValidationError: false
		}
	}
	
	static get propTypes() {
		return {
			isCreating: PropTypes.bool,
			history: PropTypes.object,
			match: PropTypes.object,
			signUpUser: PropTypes.func,
			validationModel: PropTypes.instanceOf(ValidationModel)
		}
	}

	componentDidUpdate(prevProps) {
		if(prevProps.isCreating && !this.props.isCreating) {
			this.setState({showSignUpSuccessModal: true})
		}

		if(prevProps.validationModel.isValid !== this.props.validationModel.isValid) {
			this.setState({validationModel: this.props.validationModel})
		}
	}

	_onPropertyChange = (user, validationModel) => {
		this.setState({user, validationModel})
	}

	_onSave = () => {
		const {user} = this.state;
		const {signUpUser} = this.props;
		const validationModel = validateSignUpForm(user);
		this.setState({ 
			validationModel,
			showValidationError: !validationModel.isValid
		});
		if(validationModel.isValid) {
			signUpUser(user.mapToSignUpApi());
		}
	}

	_onSignUpSuccessSave = () => {
		const { history } = this.props;
		this.setState({showLogInSuccessModal: false}, () => history.push(`/login/${true}`));
	}

	render(){
		const { match, history} = this.props;
		const {user, validationModel, showSignUpSuccessModal, showValidationError} = this.state;

		const modalProps = {
			title: "Create an Account",
			show: match?.params?.show ? true : false,
			onHide: () => history.push('/')
		}
		
		const formProps = {
			user,
			onPropertyChange: this._onPropertyChange,
			onSave: this._onSave
		}

		return (
			<Modal {...modalProps}>
				{!showSignUpSuccessModal ? (
				<>
					<ValidationSummary show={showValidationError && !validationModel.isValid} validationModel={validationModel} />
					<SignUpForm {...formProps} />
				</>
				) : (
					<SignUpSuccessForm onSave={this._onSignUpSuccessSave} />
				)}
			</Modal>
		)
	}
}

export default withRouter(connect(mapStoreToProps, mapDispatchToProps)(SignUp));