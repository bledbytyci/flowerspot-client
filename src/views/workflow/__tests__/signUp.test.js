import React from 'react';
import { SignUp } from '../signUp.jsx';
import { shallow } from 'enzyme';
import ValidationModel from '../../../setup/validationModel.js';
import toJson from 'enzyme-to-json';

jest.unmock('../signUp.jsx');
jest.unmock('../../../user/user.js');
jest.unmock('../../../setup/validationModel.js')

const shallowOptions = {
    lifecycleExperimental: true,
    disableLifecycleMethods: false
};

describe('SignUp component', () => {
    it('component is being rendered', () => {

		const signUpComponent = shallow(<SignUp />);
        expect(signUpComponent).not.toBe(null);
	});

	it('sets showSignUpSuccessModal to true after creating a new user', () => {
		const props = {
			isCreating: true,
			validationModel: new ValidationModel()
		}
		
		const signUpComponent = shallow(<SignUp {...props} />, shallowOptions);

		props.isCreating = false;

		signUpComponent.setProps(props);
        const currentState = signUpComponent.state();

        expect(currentState.showSignUpSuccessModal).toBe(true);
	});
});

describe('snapshots', () => {
	it('App snapshot', () => {
		const signUpComponent = shallow(<SignUp />);
		expect(toJson(signUpComponent)).toMatchSnapshot();
	})
})