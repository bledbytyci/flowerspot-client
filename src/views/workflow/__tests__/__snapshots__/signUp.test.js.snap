// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots App snapshot 1`] = `
<Modal
  onHide={[Function]}
  title="Create an Account"
>
  <ValidationSummary
    show={false}
    validationModel={
      ValidationModel {
        "isValid": true,
        "validationMessages": Immutable.List [],
      }
    }
  />
  <SignUpForm
    onPropertyChange={[Function]}
    onSave={[Function]}
    user={
      User {
        "date_of_birth": null,
        "email": "",
        "first_name": "",
        "id": "",
        "last_name": "",
        "password": "",
      }
    }
  />
</Modal>
`;
